MPR#7548 Add an example of using the printf function to the manual

Lereena opened the pull request:
Added a description and examples of using the printf function to the manual section "1.8 Pretty-printing", closing the [MPR#7548](https://caml.inria.fr/mantis/view.php?id=7548).

Lereena committed:
Add an example of using the printf function

mseri commented on manual/manual/tutorials/coreexamples.etex:
```diff
@@ -677,6 +677,36 @@ print_expr e; print_newline ();;
+their format, and arguments for output that correspond to the formatting
+directives in the format string.
+\begin{caml_eval}
+open Printf;;
```
I would avoid opening `Printf` and explicitly call `Printf.printf` in the three places it is used below

Octachron replied:
Or `Format.printf` .

mseri commented on manual/manual/tutorials/coreexamples.etex:
```diff
@@ -677,6 +677,36 @@ print_expr e; print_newline ();;
+open Printf;;
+\end{caml_eval}
+\begin{caml_example}{toplevel}
+printf "%i + %i is an integer value, %F * %F is a float, \"%s\"\n"
```
`\"%s\"` could be replaced by `%S`

mseri commented on manual/manual/tutorials/coreexamples.etex:
```diff
@@ -677,6 +677,36 @@ print_expr e; print_newline ();;
+\begin{caml_example}{toplevel}[error]
+printf "Float value: %F" 42;;
+\end{caml_example}
+Associated with this feature is the need to explicitly indicate that the
```
It would be nice, if possible, to internally link to the `Printf` module page and the section discussin the conversion specification and the types available (see https://caml.inria.fr/pub/docs/manual-ocaml/libref/Printf.html)

gasche replied:
This is beyond my own skills with manual references. @Octachron, do you remember whether we can link from the manual to the stdlib documentation? (We should describe how to do this in the manual README.)

Octachron replied:
It is possible, but the link target is  backend dependent. It is probably simpler that I add a latex macro to take care of those differences.

Lereena replied:
@Octachron If I use the \href tag and add a link to the specified documentation page via it, would this be an acceptable solution? Or should it be postponed?

Octachron replied:
We can start with a `href` and update it later indeed.

mseri commented on manual/manual/tutorials/coreexamples.etex:
```diff
@@ -677,6 +677,36 @@ print_expr e; print_newline ();;
+\begin{caml_example}{toplevel}
+printf str 3 4.5 "string value";;
+\end{caml_example}
+
```
I think it would be useful to also get the previous pretty print example rewritten in terms of `Format` and `Printf`, showing how to create pretty printers and use them (even if in "real-life" I would probably rely on the `fmt` library)

Octachron replied:
I concur. Having an example of a "%a" printer along the way would be nice too.

Octachron commented on manual/manual/tutorials/coreexamples.etex:
```diff
@@ -677,6 +677,36 @@ print_expr e; print_newline ();;
 print_expr (deriv e "x"); print_newline ();;
 \end{caml_example}
 
+There is a "printf" function in "Printf" module that allows you to make
```
I would rather mention the more featureful `Format` module.

More importantly, modules are only introduced in the next chapter, a forward reference would be nice. Something like
```latex
(see chapter~\ref{c:moduleexamples})
```

gasche replied:
Wouldn't it make sense to have a version with the simpler Printf first, and maybe (but not in the present PR) a mention of Format later?

Octachron replied:
This is certainly an option; but since Format features are a superset of Printf, skipping `Printf` works too .

gasche replied:
Here is why I'm not convinced by this idea:

- `Format` is more complex than `Printf`, because of formatting boxes and the @!%weird syntax they use in format strings. If we don't introduce them, Format adds no value, but if we do introduce them, they need quite a bit more introduction on their own, so much that showing Printf first does not make the whole thing much more cumbersome to read (it's going in increment of complexity).

- Printf is arguably the best choice for format-string printing in the OCaml ecosystem (Jane Street used to have its own printf-style extension, I'm not sure if it's actively used anymore?); it's a part of the language tutorial that users will certain find used in real code. On the other hand, there are quite a few more pretty-printing libraries than Format, some built on top of it, some on entirely different formatting engines, so this part of the tutorial (is nice to have but) would feel less canonical.

Octachron commented on manual/manual/tutorials/coreexamples.etex:
```diff
@@ -677,6 +677,36 @@ print_expr e; print_newline ();;
 \end{caml_example}
 
+There is a "printf" function in "Printf" module that allows you to make
+formatted output more concisely. It models the behavior of the "printf"
```
In this context, `model` sounds a bit off to my ear, would `follow` be more straigthfoward?

gasche commented:
Thanks for the patch proposal! The content that you propose is very nice, and it will be a good addition to the tutorial. 

I think that you could have a new section to introduce Printf (for example named "Printf formats"), instead of being at the end of the "Pretty-printing" section.

The original intent of the mantis ticket you reference was to include a version of the full pretty-printer using Printf. This probably requires the use of the more advanced format specifier `%a` (and changing the function to take a `out_channel` as argument for this to work):

```ocaml
let print ppf expr =
  ...
  let rec print prec ppf expr =
     ...
      | Quot(f, g) ->
          open_paren prec 2 ppf;
          Printf.fprintf ppf "%a / %a" (print 2) f (print 3) g;
          close_paren prec 2 ppf
```

This could go at the end of the new content you wrote ("Here is how to rewrite the pretty-printer using Printf.fprintf"), probably after having introduced `fprintf` before --- like `printf`, but parametrized on the output channel. But this does not need to be part of the present Pull Request: with a separate section name, I think this is good to merge even if we don't have a port of the full pretty-printer.

(Of course the feedback of @mseri is also excellent -- but on the last point, I think the pretty-printer could wait for another PR.)

mseri commented:
@gasche I fully agree that the pretty-printer example could be added in a second stage, and I quite like the idea of having this moved earlier on (and focusing on the easier Printf more than on Format).

Octachron commented on manual/manual/tutorials/coreexamples.etex:
```diff
@@ -677,6 +677,36 @@ print_expr e; print_newline ();;
+\begin{caml_example*}{toplevel}
+let str : ('a, 'b, 'c) format =
+    "%i is an integer value, %F is a float, \"%s\"\n";;
+\end{caml_example*}
```
It is probably simpler to use `format_of_string` here. But this is kind of mildly advanced use case, it could be nice to add some warning in the introduction of this paragraph .

gasche replied:
Yes, `format_of_string` is nice -- it allows to have the string literal checked as a format without having to use an explicit type annotation. If we write an annotation, which may still be useful, we could use `_ format` instead of `('a, 'b, 'c) format`, so that it is clear that we do not give the actual instantiations of type parameters.

Octachron commented on manual/manual/tutorials/coreexamples.etex:
```diff
@@ -677,6 +677,36 @@ print_expr e; print_newline ();;
+3 2 4.5 1. "this is a string";;
+\end{caml_example}
+The printf function in the OCaml language performs type checking of arguments.
+For example, if you pass it an argument of a type that does not correspond to
```
I would delete the `For example` connector here: the following sentence is explaining the previous one rather than giving an example.

Octachron commented on manual/manual/tutorials/coreexamples.etex:
```diff
@@ -677,6 +677,36 @@ print_expr e; print_newline ();;
+printf "%i + %i is an integer value, %F * %F is a float, \"%s\"\n"
+3 2 4.5 1. "this is a string";;
+\end{caml_example}
+The printf function in the OCaml language performs type checking of arguments.
```
I think it would made sense to rephrase this sentence to  `OCaml type system checks that the type of the arguments and the specifiers are compatible`, to avoid the idea that `printf` is doing any type checking on its own.

Octachron reviewed (commented):
Thanks for the contribution! I have some remarks in term of contents and emphases but this is a good start.

gasche commented:
@Lereena: when you have time, it would be nice if you could integrate the feedback you got from the various commenters into the PR (by adding more commits or amending your commit, and pushing to the remote branch again).

Some of the comments are a bit open-ended, and it's fine if some things are left for later, but at least easy/local change suggestions that you agree would improve the result should be handled. For the rest, you should do as you think is best to get a next iteration.

Lereena commented:
@gasche @mseri @Octachron Thank you very much for feedback, I will make corrections tomorrow

Lereena committed:
Creating a new section, correcting wording

Lereena committed:
Add links

Lereena committed:
Add rewritten pretty-printer

mseri commented on manual/manual/tutorials/coreexamples.etex:
```diff
@@ -706,7 +706,38 @@ let str : _ format =
+\begin{caml_example}{toplevel}
+let print_expr ppf expr =
+  let open_paren prec op_prec output =
+    if prec > op_prec then Printf.fprintf output "%S" "(" in
```
Are you sure you want `%S` instead of `%s` here and below?

Lereena replied:
Actually no, I messed up. Thanks for noticing

mseri reviewed (commented):
Nice, you also added the pretty printer example! Thanks, this is a great addition to the examples

Lereena committed:
Change %S to %s

Lereena commented:
I corrected the wording and function calls, added links and a printer. I didn’t write about the Format module, in my opinion this is so far superfluous

gasche commented on manual/manual/tutorials/coreexamples.etex:
```diff
@@ -677,6 +677,68 @@ print_expr e; print_newline ();;
+Printf.printf "%i + %i is an integer value, %F * %F is a float, %S\n"
+3 2 4.5 1. "this is a string";;
+\end{caml_example}
+OCaml type system checks that the type of the arguments and the specifiers are
```
Here it would be "The OCaml type system"

gasche commented on manual/manual/tutorials/coreexamples.etex:
```diff
@@ -677,6 +677,68 @@ print_expr e; print_newline ();;
+Printf.printf str 3 4.5 "string value";;
+\end{caml_example}
+"fprintf" function is like "printf" except that it takes an output channel as
+the first argument. Here is how to rewrite the pretty-printer using "fprintf":
```
Before the big example, it would be nice to have an easy example of fprintf in action, and then an easy example of `%a` in action. Here would be possible suggestions:

```
let output_int ppf n = Printf.fprintf ppf "%d" n

let output_option printer ppf = function
  | None -> Printf.fprintf ppf "None"
  | Some v -> Printf.fprintf ppf "Some(%a)" printer v

let () = Printf.fprintf stdout
  "The current setting is %a"
  (output_option output_int) (Some 3)
```

gasche reviewed (commented):
I made a couple minor comments but this looks almost ready for merging. Thanks!

Lereena committed:
Add an easy example of %a

Lereena commented:
@gasche I fixed wording and used an example that you proposed, thank you very much.

mseri commented on manual/manual/tutorials/coreexamples.etex:
```diff
@@ -707,7 +707,18 @@ let str : _ format =
 \end{caml_example}
 "fprintf" function is like "printf" except that it takes an output channel as
-the first argument. Here is how to rewrite the pretty-printer using "fprintf":
+the first argument.
```
I think that we should explain carefully what is going on in the preliminary example:

- explain what are you doing defining the `output_int` printer

-  then mention that this approach works for more complex types, for example options, and that in such case you also need to provide a printer for any additional type that they wrap,

- then explain that in this way you are constructing composable printers that can be used with the `%a` format specifier and conclude the example.

This will greatly clarify for the reader both the introductory examples that you just added, and the more complex `print_expr` presented afterwards

Octachron replied:
I agree with @mseri that the next example lacks a short introduction.
For instance,
> The `%a` specifier can be useful to define custom printer (for custom types). For instance, a custom printer for ints:

Then gives an example of using `pp_int` with `%a`, before introducing `pp_option` as a combinator that transforms a printer for `'a` into a printer for `'a option`.

Lereena replied:
@Octachron Sorry, but I didn't understand what kind of example is needed. Should I somehow wrap int to use it with %a?

Octachron replied:
The example with `pp_int`? I was thinking of something like:
```OCaml
\begin{caml_example}{toplevel}
let pp_int ppf n = Printf.fprintf ppf "%d" n;;
Printf.printf "Using a custom printer for integer %a\n" pp_int 12
\end{caml_example}{toplevel}
```
to show that `pp_int` is a composable printer by itself.

mseri replied:
I agree. I would also split the example in its consituents, somewhat like:
```
[...]the first argument. This is useful to construct composable
printers that can be used with the "%a" specifier.
For example, we can create a printing template that converts
an integer argument to signed decimal as follows.
\begin{caml_example}{toplevel}
let pp_int ppf n = Printf.fprintf ppf "%d" n;;
Printf.printf "Using a custom printer for integer %a\n" pp_int 12
\end{caml_example}{toplevel}
As you can see, the printing template is itself a composable
printer. We can push this example further, and define a printer
for optional types.
\begin{caml_example}{toplevel}
let pp_option printer ppf = function
  | None -> Printf.fprintf ppf "None"
  | Some v -> Printf.fprintf ppf "Some(%a)" printer v
+ couple of examples of use if you want
\end{caml_example}{toplevel}
Here you see that we need to also provide a printer as argument, 
indeed "pp_option" is a function that processes the argument
passed to it and depending on its value prints "None"  or usest
the provided printer.
[...]
```
Guiding the reader through the example, instead of showing
the example and then explaining it as a whole.

mseri replied:
Don't get me wrong, this is already a very good addition to the manual, but I think that a small restructuring can make it even better

Lereena replied:
Thank you very much. I added explanations and restructured them.

Lereena committed:
Add an explanation of an example

Lereena commented:
I am very sorry and extremely embarrassed that my contribution is so long and takes a lot of your attention D: Thank you for helping.

Octachron commented:
@Lereena , you should not feel embarrassed: it is perfectly normal to have a lot of back-and-forth between the author and editors(or reviewers) before reaching the "right" text. Clarity and simplicity sometimes requires a lot of work, and this PR is on the right track.

Octachron commented on manual/manual/tutorials/coreexamples.etex:
```diff
@@ -677,6 +677,89 @@ print_expr e; print_newline ();;
+\end{caml_example*}
+\begin{caml_example}{toplevel}
+Printf.printf str 3 4.5 "string value";;
+\end{caml_example}
```
Personally, I would move this example at the end of the section after the `print_expression` example since knowing how to store format string in variables feels less crucial than knowing how to define composable printers.

Lereena replied:
In my opinion, it is worth going from the simpler to the complex, which is why I left this example before starting the description of composable printers.

Octachron replied:
My opinion is precisely that this example is much more complex and uncommon than composable printers. In particular, the definition of `str` yields the following toplevel output:
```OCaml
# let str: _ format = "%i is an integer value, %F is a float, %S\n";;

val str : (int -> float -> string -> 'a, 'b, 'a) format =
  CamlinternalFormatBasics.Format
   (CamlinternalFormatBasics.Int (CamlinternalFormatBasics.Int_i,
     CamlinternalFormatBasics.No_padding,
     CamlinternalFormatBasics.No_precision,
     CamlinternalFormatBasics.String_literal (" is an integer value, ",
      CamlinternalFormatBasics.Float (CamlinternalFormatBasics.Float_F,
       CamlinternalFormatBasics.No_padding,
       CamlinternalFormatBasics.No_precision,
       CamlinternalFormatBasics.String_literal (" is a float, ",
        CamlinternalFormatBasics.Caml_string
         (CamlinternalFormatBasics.No_padding,
         CamlinternalFormatBasics.Char_literal ('\n',
          CamlinternalFormatBasics.End_of_format)))))),
   "%i is an integer value, %F is a float, %S\n")
```
which is quite esoteric-looking for  … anybody not familiar with the implementation of format strings. You rightfully elided this output, but the complexity is still here.

Lereena replied:
Indeed, a fair point. It is probably worthwhile to move this example to the end of the section, but leave the output hidden. For the first acquaintance with the language, it is probably enough to know about the existence of the ability to store strings in variables, but not necessarily to get acquainted with the implementation details.

Drup commented:
I like the idea of introducing printf in the tutorial! :)

I would suggest using `pp_*` as a prefix. `pp` is the standard name for pretty printers, and there is already a set of `output_*` function in `Pervasive` which do something quite different.

I'm also of the opinion we should advertise `Format` directly. Especially since in this case, the code is strictly identical. Although I understand that not everyone agrees on that.

Lereena committed:
Rearrange examples

Lereena committed:
Change prefixes

Lereena commented:
> I would suggest using `pp_*` as a prefix. `pp` is the standard name for pretty printers, and there is already a set of `output_*` function in `Pervasive` which do something quite different.

I agree, there can really be a misunderstanding. Made a change

Armael commented on manual/manual/tutorials/coreexamples.etex:
```diff
@@ -677,6 +677,90 @@ print_expr e; print_newline ();;
+
+Here is how to rewrite the pretty-printer using "fprintf":
+\begin{caml_example}{toplevel}
+let print_expr ppf expr =
```
`print_expr` could also be renamed to `pp_expr` for consistency, since it could itself be used with "%a" as a composable printer!

Armael commented on manual/manual/tutorials/coreexamples.etex:
```diff
@@ -677,6 +677,90 @@ print_expr e; print_newline ();;
+passed as an argument).
+This approach works for more complex types, for example options. In such case
+we need to provide a printer for any additional type that they wrap.
+In this way we coustruct the composable printer that can be used with the "%a"
```
coustruct -> construct

Armael reviewed (commented):
A couple of minor remarks.

Lereena committed:
Rename function and fix typo

Lereena committed:
Add a mention of the printf tutorial section to Changes

Lereena committed:
Add an introduction of custom printer for int

Lereena committed:
Restructure explanation

Octachron commented on manual/manual/tutorials/coreexamples.etex:
```diff
@@ -677,6 +677,95 @@ print_expr e; print_newline ();;
+We need to also provide a printer as argument. "pp_option" processes the
+argument depending on its value by printing "None" or using the provided printer.
+This approach works for more complex types like options. In such case
+we need to provide a printer for any additional type that they wrap.
```
> This approach works ... that they wrap.

This is now redundant with the introduction, you should remove these two sentences..

Octachron commented on manual/manual/tutorials/coreexamples.etex:
```diff
@@ -677,6 +677,95 @@ print_expr e; print_newline ();;
+print_newline();;
+Printf.fprintf stdout "There is only %a"
+  (pp_option pp_int) (None);
+print_newline();;
```
I think that the example could be simplified to

```OCaml
Printf.fprintf stdout
  "The current setting is %a\n\
    There is only %a\n"
  (pp_option pp_int) (Some 3)
  (pp_option pp_int) None
```
At least, removing the calls to `print_newline` would be nice.

Octachron commented on manual/manual/tutorials/coreexamples.etex:
```diff
@@ -677,6 +677,95 @@ print_expr e; print_newline ();;
+  (pp_option pp_int) (None);
+print_newline();;
+\end{caml_example}
+We need to also provide a printer as argument. "pp_option" processes the
```
`"pp_option"` → `The function "pp_option"`?
Personally, I prefer to not start sentence with identifier names

Lereena committed:
Fix an example

Octachron commented on Changes:
```diff
@@ -465,6 +465,10 @@ OCaml 4.08.0
 ### Manual and documentation:
 
+- MPR#7548: printf example in the tutorial part of the manual
+ (Kostikova Oxana, rewiew by Gabriel Scherer, Florian Angeletti, 
```
There is whitespace at the end of this line that should be removed to make check-typo happy.

Octachron commented on manual/manual/tutorials/coreexamples.etex:
```diff
@@ -677,6 +677,93 @@ print_expr e; print_newline ();;
+Printf.printf "Outputting an integer using a custom printer: %a " pp_int 42;;
+\end{caml_example}
+The printing template is itself a composable printer. It is also possible
+to create a more complex printer, for example for the optional types:
```
`for the optional types` → `for optional types`. I would also transform the comma into a dot (shorter sentences tends to work better in English technical writings) :

> to create more complex printers. For example, for optional types

Octachron commented on manual/manual/tutorials/coreexamples.etex:
```diff
@@ -677,6 +677,93 @@ print_expr e; print_newline ();;
+let pp_int ppf n = Printf.fprintf ppf "%d" n;;
+Printf.printf "Outputting an integer using a custom printer: %a " pp_int 42;;
+\end{caml_example}
+The printing template is itself a composable printer. It is also possible
```
At this point, the reader has not been introduced to the notion of `composable printer` yet. I think it would work better to tell the reader that `pp_int` is a composable reader and that those printers can be used to build printers for more complex types. For instance, we could merge the two sentences into something like:

> The advantage of those printers based on the "%a" specifier is that they can be composed together to create more complex printers step by step.

Octachron commented on manual/manual/tutorials/coreexamples.etex:
```diff
@@ -677,6 +677,93 @@ print_expr e; print_newline ();;
+\end{caml_example}
+The printing template is itself a composable printer. It is also possible
+to create a more complex printer, for example for the optional types:
+"'a" type into a printer for "'a option":
```
There are some missing words here:
> "'a" type into a printer for "'a option":

Maybe

>  we can define a combinator that can turn a printer for a type "'a" into a printer for "'a optional"

?

Octachron commented on manual/manual/tutorials/coreexamples.etex:
```diff
@@ -677,6 +677,93 @@ print_expr e; print_newline ();;
+  (pp_option pp_int) None
+;;
+\end{caml_example}
+We need to also provide a printer as argument. The function "pp_option"
```
I am not sure what the `also` is referring to here. I think this first sentence can be removed.

Octachron commented on manual/manual/tutorials/coreexamples.etex:
```diff
@@ -677,6 +677,93 @@ print_expr e; print_newline ();;
+\end{caml_example}
+We need to also provide a printer as argument. The function "pp_option"
+processes the argument depending on its value by printing "None" or using
+the provided printer.
```
`process the argument` and `depending of its value` seem redundant with each other. Maybe

> If the value of its argument its "None", the printer returned by `pp_option printer` prints "None" otherwise it uses the provided printer to print "Some <value>".

But I am not sure if the code of the function need to be explained.

Octachron commented on manual/manual/tutorials/coreexamples.etex:
```diff
@@ -677,6 +677,93 @@ print_expr e; print_newline ();;
+    if prec > op_prec then Printf.fprintf output "%s" ")" in
+  let rec print prec ppf expr =
+      match expr with
+        Const c -> Printf.fprintf ppf "%F" c
```
Optional nitpick: `| Const c` is more in line with current style.

Octachron commented on manual/manual/tutorials/coreexamples.etex:
```diff
@@ -677,6 +677,93 @@ print_expr e; print_newline ();;
+\end{caml_example}
+
+Associated with typechecking feature is the need to explicitly indicate that the
+string is a format string in order to pass it to the "printf" function.
```
I think this sentence could be reworded to make it clearer that this is only the case if one need to store the format string. For instance,

>  Due to the way that format string are build, storing a format string requires an explicit type annotation:
\begin{caml_example*}{toplevel}
...

Similarly, we could explain after the example that without the annotation `str` would be a string.

> Without this annotation, `str` would be a string and not a format string.

Octachron commented on manual/manual/tutorials/coreexamples.etex:
```diff
@@ -677,6 +677,93 @@ print_expr e; print_newline ();;
+\begin{caml_example}{toplevel}[error]
+Printf.printf "Float value: %F" 42;;
+\end{caml_example}
+"fprintf" function is like "printf" except that it takes an output channel as
```
There is a missing article here: '*The* "fprintf" function'

Octachron reviewed (commented):
I have done one more round of review. The new section  shaping up nicely and it should be good to go after that round.

Lereena committed:
Wording and typos correction

Octachron commented on manual/manual/tutorials/coreexamples.etex:
```diff
@@ -677,6 +677,93 @@ print_expr e; print_newline ();;
+It follows the behavior of the "printf" function from the C standard library.
+The "printf" function takes a format string that describes the output values and
+their format, and arguments for output that correspond to the formatting
+directives in the format string.
```
I have one last comment before approving and merging: I fear that the above sentence

> The "printf" function takes a format string that describes the output values ... in the format string.

 does not not work. Reading the sentence, I am left with four questions
- what are those `output values`?
- what are their format ?
- what is the difference between `output values` and `argument for output`?
- what are formatting directives?

On the first two questions, I think that you are trying to describe the specifiers, but then the proposition

> a format string that describes the output values and their format

is problematic because the format string is more than a collection of specifiers: `%a%a` ought to be different from `"%a +%a"`. Similarly, the last two questions arises because  the connection between the specifiers and the arguments is muddled by the apparition of a new term `formatting directives` which was never mentioned before.

 I think you meant something like

> The first argument of this "printf" function is a format string that describes the desired output as a text interspersed with specifiers (for instance "%d", "%f", \ldots) . The following arguments are then used to fill the holes left by the specifiers inside the format string:

Like always, feel free to reformulate the above text or even discard it, but keep in mind that this is an introduction, new terms should be as sparse as possible.

Lereena replied:
In the last commit I tried to simplify as much as possible

Lereena committed:
Simplify paragraph

Octachron commented on manual/manual/tutorials/coreexamples.etex:
```diff
@@ -677,6 +677,94 @@ print_expr e; print_newline ();;
+module (see chapter~\ref{c:moduleexamples}) that allows you to make formatted
+output more concisely.
+It follows the behavior of the "printf" function from the C standard library.
+The "printf" function takes a format string that describes the desired output
```
Would `takes as argument`  be clearer?

Lereena replied:
I don't think so. This is a fairly well-established expression, hardly anyone would think that the function takes it somehow differently.

Octachron replied:
Ok.

Octachron commented on manual/manual/tutorials/coreexamples.etex:
```diff
@@ -677,6 +677,94 @@ print_expr e; print_newline ();;
+It follows the behavior of the "printf" function from the C standard library.
+The "printf" function takes a format string that describes the desired output
+as a text interspered with specifiers (for instance "%d", "%f").
+Next, the specifiers are substituted by the corresponding arguments in the
```
The term `corresponding arguments` may be unclear. A reader might think that it mean that

```OCaml
Printf.printf "%d + %B" true 1 
```
works because `true` is the corresponding argument for `%B`.
Maybe
> the following arguments in their order of apparition

?

Octachron reviewed (commented):
We are near here! I have two small remaining remarks:

Lereena committed:
Clarification

Octachron reviewed (approved):
I like the last version. I will merge it once CI passes. Thanks for your patience!
